Issue/162/unit tests with physical values

m-aguena opened the pull request:
added `NumCosmo` values to unit tests of modelling

m-aguena commented:
@cavestruz @matthewkirby @aimalz Right now the unit tests of modeling are failing because the precision of `get_critical_surface_density` in Travis is 2e-4 comparing to the value I got with `NumCosmo`. However when I compare with the value I get in my pc and in the binder @vitenti made I have a precision of 1e-8. I am using the same astropy version (v3.2.3) as Travis so I don't know where this difference came from. I don't know the best approach to proceed, do you think this is something to be investigated or just decrease the precision used in the unit tests?

matthewkirby commented:
Sounds like a bug to me. We should track it down at least why it’s
different. That function of ours you listed isn’t actually tested in or
unit tests. So it may be wrong. Check the values of the speed of light and
Newton’s constant between the versions as well because that could also be
the source.

On Wed, Nov 20, 2019 at 7:21 PM Michel Aguena <notifications@github.com>
wrote:

> @cavestruz <https://github.com/cavestruz> @matthewkirby
> <https://github.com/matthewkirby> @aimalz <https://github.com/aimalz>
> Right now the unit tests of modeling are failing because the precision of
> get_critical_surface_density in Travis is 2e-4 comparing to the value I
> got with NumCosmo. However when I compare with the value I get in my pc
> and in the binder @vitenti <https://github.com/vitenti> made I have a
> precision of 1e-8. I am using the same astropy version (v3.2.3) as Travis
> so I don't know where this difference came from. I don't know the best
> approach to proceed, do you think this is something to be investigated or
> just decrease the precision used in the unit tests?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/LSSTDESC/CLMM/pull/188?email_source=notifications&email_token=ABJFFNMR2NDTVRXTN54K2NLQUXOPDA5CNFSM4JP2XRZ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEXVR2A#issuecomment-556751080>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABJFFNMEO27Z3C2WKVLMSZLQUXOPDANCNFSM4JP2XRZQ>
> .
>

matthewkirby commented:
When this is ready for a review, tag it up as such. I want to note that my comment from your other open PR applies here. It is really really bad practice to use a pickle file like this. You can't see inside and there is no limitation to what that code could do. For instance, it could install a bunch of dependencies and then malware and you can't know for sure until you open it and try. There really is no benefit to saving the numcosmo results in a pickle file over a text file so lets just use a text file.

cavestruz commented:
Quick question -  I noticed there's an rtol defined but no atol @m-aguena .  When Shenming and I were looking at the test_polaraveraging.py functions, both atol and rtol needed to be non-zero to account for certain cases.  I haven't carefully looked at the code yet, but this could be leading to the/a bug.  Happy to jump in on this first thing tomorrow and pair-code if that is helpful.

m-aguena commented:
> When this is ready for a review, tag it up as such. I want to note that my comment from your other open PR applies here. It is really really bad practice to use a pickle file like this. You can't see inside and there is no limitation to what that code could do. For instance, it could install a bunch of dependencies and then malware and you can't know for sure until you open it and try. There really is no benefit to saving the numcosmo results in a pickle file over a text file so lets just use a text file.

I just removed the picked file and put all saved values to txt file

m-aguena commented:
I added some tests where the computation of sigma_crit was done explicitly in the test function and got the same result. Than I noticed that the constants in master changed. They should have the same values but somehow because they were defined in different units and than converted there might be a difference there. I will investigate but it was either
```
# old
clight_pc_s = constants.c.to(units.pc/units.s).value
# new
CLIGHT_KMS = astropyconst.c.to(u.km/u.s).value
clight_pc_s = const.CLIGHT_KMS.value * 1000. / const.PC_TO_METER.value
 ```
or
```
# old
gnewt_pc3_msun_s2 = constants.G.to(units.pc**3/units.M_sun/units.s**2).value
# new
GNEWT = astropyconst.G.to(u.m**3/u.kg/u.s**2).value
gnewt_pc3_msun_s2 = const.GNEWT.value * const.SOLAR_MASS.value / const.PC_TO_METER.value**3
```

m-aguena commented:
Finally found the problem, the `SOALR_MASS` constant didn't have enough decimal places. I updated it from `1.98892e30` to `1.9884754153381438e+30` and the tests passed. A good example of the importance of test with actual physical values.
@cavestruz @aimalz @matthewkirby

m-aguena commented:
I reverted the value of solar mass to what was defined before based on ccl and added a conversion in the unit tests to correct for the difference in the physical values.

cavestruz commented:
@m-aguena - I just received a request for review.  Just checking, is this "ready for review" (in which case, can you tag "PR: ready for review") or was I just added for the future when it is ready (in which case, can you tag "PR: not ready for review")?  Thanks for pushing this forward!

m-aguena commented:
> @m-aguena - I just received a request for review. Just checking, is this "ready for review" (in which case, can you tag "PR: ready for review") or was I just added for the future when it is ready (in which case, can you tag "PR: not ready for review")? Thanks for pushing this forward!

I didn't know we were using that tag. Added it now

matthewkirby commented on tests/test_modeling_physical_validation.py:
```diff
@@ -0,0 +1,161 @@
+
+# Read test case
+CODE_PATH = '/'.join(os.path.abspath(__file__).split('/')[:-2])
+VAL_FILES_PATH = '%s/tests/physical_values_nc'%CODE_PATH
```
We shouldn't need to add anything to path so relative path work. `pytest` runs from the root package directory.

m-aguena replied:
fixed

matthewkirby commented on tests/test_modeling_physical_validation.py:
```diff
@@ -0,0 +1,161 @@
+VAL_FILES_PATH = '%s/tests/physical_values_nc'%CODE_PATH
+with open('%s/float_vals.txt'%VAL_FILES_PATH, 'r') as f:
+    TEST_CASE = ast.literal_eval(f.read())
+f.close()
```
This shouldn't be global. Lets put this into a function that our tests can call.

m-aguena replied:
fixed

matthewkirby commented on tests/test_modeling_physical_validation.py:
```diff
@@ -0,0 +1,161 @@
+NC_DIST = np.genfromtxt('%s/numcosmo_angular_diameter_distance.txt'%
+                        VAL_FILES_PATH, names=True)
+
+R3D = np.array(NC_PROF['r3d'])
```
Same as last comment. Lets put this into a function (along with everything that reads in from files).

m-aguena replied:
fixed

matthewkirby commented on tests/test_modeling_physical_validation.py:
```diff
@@ -0,0 +1,161 @@
+COSMO_APY = astropy.cosmology.core.FlatLambdaCDM(H0=TEST_CASE['cosmo_H0'],
+                                                 Om0=TEST_CASE['cosmo_Om0'],
+                                                 Ob0=TEST_CASE['cosmo_Ob0'])
+COSMO_CCL = clmm.cclify_astropy_cosmo(COSMO_APY)
```
Lets move this into functions

m-aguena replied:
fixed

matthewkirby commented on tests/test_modeling_physical_validation.py:
```diff
@@ -0,0 +1,161 @@
+    'delta_mdef':TEST_CASE['mass_Delta'],
+    'halo_profile_model':TEST_CASE['density_profile_parametrization'],
+    'z_src_model':'single_plane',
+    }
```
Move into function. We can just write one function that returns all of this stuff that each test calls.

m-aguena replied:
fixed

matthewkirby commented on tests/test_modeling_physical_validation.py:
```diff
@@ -0,0 +1,161 @@
+    Notes
+    -----
+        The precision set for these tests are somewhat arbitrary
+        has to be better defined at some point
```
This is not correct. numbers are directly from CCL. The only thing that isn't 100% identical to CCL is GNEWT. Once we have CCL integration, we will pull values from there.

m-aguena replied:
What I meant is the precision I set there were somewhat arbitrary.

matthewkirby replied:
Ahhh gotcha.

matthewkirby commented on tests/test_modeling_physical_validation.py:
```diff
@@ -0,0 +1,161 @@
+    ds_clmm = clmm.modeling.get_angular_diameter_distance_a(COSMO_CCL,
+                                                            TEST_CASE['aexp_source'])
+    dsl_clmm = clmm.modeling.get_angular_diameter_distance_a(COSMO_CCL,
+                                                             TEST_CASE['aexp_source'],
```
When the import scheme is changed above (`from clmm.modeling import get_angular_diameter_distance_a, otherfunc2, otherfunc3`) this line will be much shorter and we should be able to get the assert statements on the same lines as the computations.

matthewkirby replied:
We want this because if a test fails, we want the error message to include the function call that failed, not just `dl_clmm`

m-aguena replied:
fixed

matthewkirby commented on tests/test_modeling_physical_validation.py:
```diff
@@ -0,0 +1,161 @@
+    test_delta_sigma()
+    test_gammat()
+    test_kappa()
+    test_gt()
```
Shouldn't have a main funciton like this.

m-aguena replied:
fixed

m-aguena commented on tests/test_modeling_physical_validation.py:
```diff
@@ -0,0 +1,161 @@
+from numpy import testing as tst
+import numpy as np
+import clmm
+from clmm.constants import Constants as clmmconst
```
fixed

m-aguena commented:
> A bunch of things.
> 
> 1. There are a lot of small changes that are beyond the scope of this PR. the single line changes from `modeling.py` and `test_polaraveraging.py` for instance. They are done so let's just leave them.
> 
> * [x]  Move the numcosmo data files into the `tests/data` directory. We can either make a directory `tests/data/numcosmo/`, put them in there, and change names to `angular_diam...etc`   or just put them in `tests/data/` with their current names `numcosmo_angular...etc`.
> * [x]  Recover `tests/test_modeling.py`. This one is important. If you delete the file like this, when I go to merge the refactored tests, it will just delete the file. Just recover it and leave it untouched. Once this is merged into master, I will manually put the contents of `test_modeling_physical_validation.py` into the appropriate places in `test_modeling.py` (before merging my PR that writes these tests).
> * [x]  Lets be a bit more descriptive with the names of the numcosmo files. `float_vals.txt` for instance. Maybe this can be `numcosmo_config.txt` or something?
> * [x]  See inline comments for `test_modeling_physical_validation.py`. Generally, all of the code should be in a function. We also shouldn't have a main function in the test files. If you ever want to run a specific test only you can run `pytest path/to/test/file.py`.
> * [x]  This is nitpicky but docstrings should be 3x double quotes `"""` not single quotes `'''`. Lets just find and replace this before merging it in to be consistent across the project.
> * [x]  We should aim to have each assert statement have the function call within it rather than saving to a variable and then asserting.
> * [ ]  Lets also make a global variable `TOLERANCE` that has an rtol and atol value used for all of the tests rather than seemingly pulling different tolerances out of a hat for each test. This should be consistent with the rest of the test files.
> * [ ]  All notebooks in the examples directory should be examples that anyone who comes along to use the code can run. This is not true for this new notebook. Also, it isn't an example. If we have replaced the contents of this with automated validation tests, we no longer need this notebook.

@matthewkirby 

About the `TOLERANCE`, how should we deal if I know different components will have different agreement levels with NumCosmo? Should we level by the lowest tolerance?

About the notebook, I agree it does not have the same functionality as the other ones, and the comparisons in there were passed to the unit tests in `tests/test_modeling_physical_validation.py`. However I think it is very useful as it provides a guideline to compute the same properties with CLMM and NumCosmo. It probably has to be renamed and maybe moved somewhere else? (where I don't know though). Anyhow I think doing this should be in the scope of Issue #178

matthewkirby commented:
> About the TOLERANCE, how should we deal if I know different components will have different agreement levels with NumCosmo? Should we level by the lowest tolerance?

Good point, these tests are not the same as the other. Doesn't make as much sense here.

> About the notebook, I agree it does not have the same functionality as the other ones, and the comparisons in there were passed to the unit tests in tests/test_modeling_physical_validation.py. However I think it is very useful as it provides a guideline to compute the same properties with CLMM and NumCosmo. It probably has to be renamed and maybe moved somewhere else? (where I don't know though). Anyhow I think doing this should be in the scope of Issue #178

Alright, we can figure that out there then.

matthewkirby commented:
Assuming no one else has anything to say, I am happy merging this.

m-aguena commented:
@matthewkirby I looks like everything is still working after merging master that had #173 in it. Would you like to merge it so that you can move on with #200 ?

matthewkirby commented:
Lets do it!
